File: iosApp/iosApp/Features/Addons/Tableport/StampImagePrefetcher.swift
import Foundation
import Shared

/// Warms `URLCache.shared` with stamp PNGs so the next page's images are
/// already on disk by the time the user flips to it.
///
/// Strategy: fire one URLSession data task per image, low priority,
/// rate-limited so we don't saturate the network. Each request reuses
/// `URLCache.shared` (configured at app boot in `CachedImageCacheConfig`).
enum StampImagePrefetcher {
    private static let queue = DispatchQueue(label: "mensa.stamp.prefetch", qos: .utility)
    private static var inflight: Set<String> = []
    private static let session: URLSession = {
        let cfg = URLSessionConfiguration.default
        cfg.requestCachePolicy = .returnCacheDataElseLoad
        cfg.urlCache = URLCache.shared
        cfg.httpMaximumConnectionsPerHost = 4
        return URLSession(configuration: cfg)
    }()

    /// Scalda solo la pagina attorno a `page`, non tutta la collezione.
    ///
    /// `warmAll` accodava una richiesta per ogni timbro. La sessione tiene 4
    /// connessioni per host, quindi le immagini della pagina che l'utente sta
    /// *guardando* finivano in fila dietro decine di prefetch di pagine mai
    /// aperte: e' esattamente il caricamento in ritardo che si vede a schermo.
    /// Una pagina avanti e una indietro bastano, il libro si sfoglia una
    /// pagina per volta.
    static func warmAround(_ stamps: [StampUserModel], page: Int, perPage: Int) {
        guard !stamps.isEmpty, perPage > 0 else { return }
        let from = max(0, (page - 1) * perPage)
        let to = min(stamps.count, (page + 2) * perPage)
        guard from < to else { return }
        warm(Array(stamps[from..<to]))
    }

    private static func warm(_ stamps: [StampUserModel]) {
        queue.async {
            for s in stamps {
                guard let url = url(for: s) else { continue }
                let key = url.absoluteString
                if inflight.contains(key) { continue }
                inflight.insert(key)
                var req = URLRequest(url: url, cachePolicy: .returnCacheDataElseLoad)
                req.timeoutInterval = 20
                let task = session.dataTask(with: req) { _, _, _ in
                    queue.async { inflight.remove(key) }
                }
                task.priority = URLSessionTask.lowPriority
                task.resume()
            }
        }
    }

    private static func url(for stamp: StampUserModel) -> URL? {
        guard let r = stamp.stampRecord, !r.image.isEmpty else { return nil }
        return Files.url(collection: "stamp", recordId: r.id, filename: r.image, thumb: PassportLayout.gridThumb)
    }
}

File: androidApp/src/main/kotlin/it/mensa/app/features/tableport/util/StampImagePrefetcher.kt
package it.mensa.app.features.tableport.util

import android.content.Context
import coil3.ImageLoader
import coil3.request.ImageRequest
import it.mensa.app.features.tableport._components.StampThumb
import it.mensa.app.support.FilesUrl
import it.mensa.shared.model.StampUserModel

/**
 * StampImagePrefetcher — pre-fetches stamp images into the Coil disk cache.
 * Mirrors iOS StampImagePrefetcher.swift.
 *
 * Call [warmAll] after stamps are loaded to ensure images are cached before
 * the user scrolls through passport pages.
 */
object StampImagePrefetcher {

    /**
     * Scalda solo la pagina attorno a [pageIndex], non tutta la collezione.
     *
     * `warmAll` accodava ogni timbro in una volta sola. Coil serve le
     * richieste da un pool limitato, quindi le immagini della pagina che
     * l'utente sta *guardando* finivano in coda dietro decine di prefetch di
     * pagine che non aveva ancora aperto: il risultato e' proprio il
     * caricamento in ritardo di quello che si vede.
     *
     * Una pagina avanti e una indietro bastano: il pager si sfoglia una
     * pagina per volta.
     */
    fun warmAround(
        stamps: List<StampUserModel>,
        pageIndex: Int,
        perPage: Int,
        context: Context,
        imageLoader: ImageLoader,
    ) {
        if (stamps.isEmpty() || perPage <= 0) return
        val from = ((pageIndex - 1) * perPage).coerceAtLeast(0)
        val to = ((pageIndex + 2) * perPage).coerceAtMost(stamps.size)
        if (from >= to) return
        warm(stamps.subList(from, to), context, imageLoader)
    }

    private fun warm(stamps: List<StampUserModel>, context: Context, imageLoader: ImageLoader) {
        stamps.forEach { su ->
            val record = su.stampRecord ?: return@forEach
            if (record.image.isEmpty()) return@forEach
            val url = FilesUrl.build(
                collection = "stamp",
                recordId = record.id,
                filename = record.image,
                thumb = StampThumb.GRID,
            )
            val request = ImageRequest.Builder(context)
                .data(url)
                .memoryCacheKey(url)
                .diskCacheKey(url)
                .build()
            imageLoader.enqueue(request)
        }
    }
}
